lib/galaxy/webapps/galaxy/controllers/history.py

```diff
@@ -969,7 +969,13 @@ def _populate_restricted(self, trans, user, histories, send_to_users, action, se
                 else:
                     # Only deal with datasets that have not been purged
                     for hda in history.activatable_datasets:
-                        if trans.app.security_agent.can_access_dataset(send_to_user.all_roles(), hda.dataset):
+                        if len(hda.dataset.storage_media_associations) > 0:
```

Comment:
This is no sufficient to prevent sharing at all I don't think. There are other paths to share datasets that don't hit this controller, library datasets should be prohibited from being such datasets, etc...

I think https://github.com/galaxyproject/galaxy/pull/10840 is what we want. It is much more general - it allows any objectstore to be marked as private - and it is much more comprehensive in how it prevents sharing. It has test cases, it prevents such datasets from even showing up where say importing history datasets into libraries, etc...

I think this portion of the PR should be dropped when that other PR is merged and instead just ensure that your user based objectstores are marked as private - I think better APIs and UIs will pretty cleanly fallout from that.